test/unit/Feedback.test.js

```diff
@@ -59,10 +61,170 @@ describe('Feedback', () => {
+  });
+
+  describe('entering text', () => {
+    it('should record text typed into form', () => {
+      const textarea = component.find('textarea');
+      textarea.instance().value = 'Test text';
+      textarea.simulate('change');
+      expect(textarea.text()).to.equal('Test text');
+    });
+  });
+
+  describe('submitting form', () => {
+
+    let savedBaseUrl;
+    let savedSetState;
+    let apiCalled;
+
+    after(() => {
+      appConfig.baseUrl = savedBaseUrl;
+    });
+
+    it('should submit form when submit is pressed', () => {
+      savedBaseUrl = appConfig.baseUrl;
+      appConfig.baseUrl = 'http://test-server.com';
+
+      return new Promise((resolve) => {
+        savedSetState = component.instance().setState.bind(component.instance());
+        component.instance().setState = (...args) => {
+          savedSetState(...args, () => { resolve(); });
+        };
+
+        nock('http://test-server.com')
+          .defaultReplyHeaders({
+            'access-control-allow-origin': '*',
+            'access-control-allow-credentials': 'true',
+          })
+          .post(/\/api\/feedback/)
+          .reply(200, () => {
+            apiCalled = true;
+            return {};
+          });
+
+        const submitButton = component.find('Button').at(2).find('button');
+        submitButton.simulate('submit');
+      }).then(() => {
+        expect(apiCalled).to.equal(true);
+      });
+    });
+
+    it('should not show We are here to help message or form', () => {
+      component.update();
+      const form = component.find('form');
+      const message = component.html();
+      expect(form.length).to.equal(0);
+      expect(message).not.to.include('We are here to help');
+    });
+
+    it('should show thank you message', () => {
+      const ptag = component.find('p');
+      const expectedText = 'Thank you for submitting your comments. ' +
+      'If you requested a response, our service staff ' +
+      'will get back to you as soon as possible.';
+      expect(ptag.text()).to.equal(expectedText);
+    });
+  });
+
+  describe('handling form failure in case of response with errors', () => {
+    component = mount(<Feedback />);
+    const textarea = component.find('textarea');
+    textarea.instance().value = 'Test text';
+    textarea.simulate('change');
+    const submitButton = component.find('Button').at(2);
+    const originalError = console.error;
+    let erroredCorrectly = false;
+
+
+    let savedBaseUrl;
+
+    after(() => {
+      appConfig.baseUrl = savedBaseUrl;
+      console.error = originalError;
+    });
+
+    it('should log the error', () => {
+      savedBaseUrl = appConfig.baseUrl;
+      appConfig.baseUrl = 'http://test-server.com';
+
+      return new Promise((resolve) => {
+
+        console.error = (...args) => {
+          originalError(...args);
+          if (args.includes('errorText')) {
+            erroredCorrectly = true;
+            resolve();
+          }
+        };
```

Comment:
I see that you're doing two things here in wrapping `console.error`: 1.  establish whether or not it was called (via `erroredCorrectly`), and 2. leveraging `console.error` calls to know when to `resolve` (i.e. to know when to call `expect`). The former is handled well by `sinon.spy`, which would allow you to assert `console.error.calledWith("errorText")` for example. For the latter, I'm not sure what to tap into. (And that may invalidate this comment.) I wonder if there's a more direct way of tapping into comonent rendering to determine when to make the test assertion.. Do you know of a way to listen on [enzyme] component updates? You wrap `setState` for this purpose elsewhere, but that's essentially what you're doing with `console.error` here.. I mean this method is fine, but it's easy to forget to restore a wrapped function, for one..

Reply:
I think this (combining checking that a method was called with special behavior) is supposed to be something you can do with `sinon.mock`, but I wasn't able to get it to work.